### ISSUE TITLE
Misleading error message when using qualified column references in UPDATE .. SET statement

### ISSUE THREAD

**Issue by lukaseder** [cockroachdb/cockroach#42362]
**Describe the problem**

When using qualified column references in `UPDATE .. SET`, CockroachDB reports a misleading error message

**To Reproduce**

Like in native PostgreSQL, the following is not supported in CockroachDB:

```sql
create table t (a int);
update t set t.a = 1;
```

While PostgreSQL reports this error:

```
SQL-Fehler [42703]: ERROR: column "t" of relation "t" does not exist
  Position: 16
```

... CockroachDB reports a more misleading one:

```
SQL-Fehler [0A000]: ERROR: at or near "a": syntax error: unimplemented: this syntax
  Detail: source SQL:
update t set t.a = 1
               ^
  Hinweis: You have attempted to use a feature that is not yet implemented.
See: https://github.com/cockroachdb/cockroach/issues/27792
```

Specifically, issue https://github.com/cockroachdb/cockroach/issues/27792 has nothing to do with this.

**Expected behavior**

Revert to PostgreSQL's native syntax error.

**Environment:**
 - CockroachDB version: CockroachDB CCL v19.2.0-rc.4 (x86_64-unknown-linux-gnu, built 2019/11/05 20:40:29, go1.12.12)
 - Server OS: Linux in Docker
 - Client app: JDBC

**Comment by awoods187** [comment]
@knz i believe you have looked at this most recently

**Comment by knz** [comment]
@lukaseder thank you for your report.

Before we proceed, can you explain a bit what is your use case and how did you notice the different error? (And how does it matter for your application?)

Also, before we proceed, I think the issue needs clarification.

You happen to be mistaken in asserting that the message is unrelated to issue  #27792.

Let me explain: in the pg syntax `UPDATE SET t.a = 123`, the text "t.a" means *"take the column called `t`, which must be of record/composite type, then set its sub-field called `a` to the value 123"*.

So there are two different things happening:

1. the column must exist
2. and also, the column must be of composite type.

To understand point 2, see the following (also using PostgreSQL):

```
kena=> create table t(t int);
CREATE TABLE
kena=> update t set t.a = 123;
ERROR:  cannot assign to field "a" of column "t" because its type integer is not a composite type
LINE 1: update t set t.a = 123;
```

It happens to fail in your example because the column `t` does not exist. But for non-existent columns, CockroachDB does indeed report the same error as postgres:

```
root@127.0.0.1:20280/defaultdb> create table t(t int);
CREATE TABLE
root@127.0.0.1:20280/defaultdb> update t set a = 123;
pq: column "a" does not exist
```

So really your example has two separate problems:

A. the column does not exist. This must result in an error.
B. CockroachdB does not support composite types. This must also result in an error.

The question you may ask is "why does CockroachDB *prefer* to report the error about composite types before the error on the missing column"?

(Indeed, pg orders this error preference differently.)

**This is because in pg, the composite type check occurs after the column existence check, whereas in crdb it occurs before.**

Sure we could change the order of the two checks but it's not exactly trivial. For us to motivate this change it would help if you could clarify what is the impact on your side.

**Comment by lukaseder** [comment]
> Before we proceed, can you explain a bit what is your use case and how did you notice the different error? (And how does it matter for your application?)

We're integrating CockroachDB support in https://jooq.org and have noticed this minor issue, which I thought was worth reporting. To jOOQ it does not matter. If a dialect does not support qualifying `SET` clause target column references, we'll simply omit the qualification in the `SET` clause generated by jOOQ

In some dialects, qualification is useful (although not necessary) when using `UPDATE .. FROM` or `UPDATE .. JOIN` clauses

> You happen to be mistaken in asserting that the message is unrelated to issue #27792.

Well, from a user perspective, I found that quite surprising :) But I can see how this may be difficult to fix.

**Comment by knz** [comment]
> In some dialects, qualification is useful (although not necessary) when using UPDATE .. FROM or UPDATE .. JOIN clauses

My current understanding:

- Even in pg, the *left* hand side of the assignments in the SET clause cannot refer to tables/columns from the secondary FROM/JOIN tables, only that of the target table. Therefore, even in pg it's not possible to qualify column names to use a different table name in the LHS. That's why pg "knows" that anything that looks like `x.y.z` must refer to a _column_ named `x` with `.y.x` part of the record type.

- The RHS can refer to other tables from the FROM/JOIN clause and there qualification is possible.

Therefore, I don't think that the  pg dialect you already have in jOOQ would qualify anything in the LHS.

If my understanding is wrong please let me know.

**Comment by lukaseder** [comment]
> Therefore, I don't think that the pg dialect you already have in jOOQ would qualify anything in the LHS. The same rules apply to crdb.

True. I was not referring to PostgreSQL dialect families specifically.

When a dialect does support full qualification, there is still a limitation where the qualification must match the target table, but when there are several conflicting column names in source and target tables, it may help readability when qualifying the LHS columns.

By default, jOOQ's integration tests try to qualify the LHS in some of the UPDATE statements, which is how I discovered this error message I've reported.

**Comment by knz** [comment]
Out of curiosity, did you choose to start a dialect from scratch instead of starting off from pg?

**Comment by lukaseder** [comment]
It's mostly starting from pg, but we also try other syntax on each new dialect integration to check what's possible, empirically.

**Comment by jasobrown** [comment]
We had a user run into this problem yesterday, using jOOQ and CockroachDB. @lukaseder is there any ticket on the jOOQ side to track any updates?

**Comment by lukaseder** [comment]
@jasobrown No there isn't. What tracking info are you looking for? This issue here is just about a misleading error message, not really about the missing feature itself...

**Comment by jasobrown** [comment]
@lukaseder It was unclear to me from this conversation if there was any work to be done on the jOOQ side to not qualify LHS arguments in an `update`, for either PG or CockrachDB integrations.

**Comment by lukaseder** [comment]
@jasobrown I'm not aware of such a problem in jOOQ. I'm very happy to review this if your team can provide a bug report: https://github.com/jOOQ/jOOQ/issues/new/choose
